### PR TITLE
Don't create two access keys for device flow login

### PR DESCRIPTION
**Pull request by jmorganca**
This PR removes the intermediate access key created during a "headless" device flow authentication. The goal is to lower our "access key sprawl".

Instead, an access key is only created in the "last leg" of the device flow process. When a user approves a device flow request, Infra track's the ID of the currently signed in user to re-create a new access key with the same fields such as provider ID.

Left to do:
- [x] Add an access key scope for approving device flow requests. That way, a key resulting from the device flow won't be able continually approve new keys (extending the expiration indefinitely).

### CODE REVIEW

File: internal/server/models/deviceflowauthrequest.go
Comment (BruceMacD) [inline internal/server/models/deviceflowauthrequest.go:16]:
```suggestion
	// ApprovedWith is the ID of the access key that approved this request
```

This looks alarming otherwise.

File: internal/server/deviceflow.go
Comment (BruceMacD) [inline internal/server/deviceflow.go:118]: nit: might make more sense to do this before creating the access key. The end result is the same in both cases though where we would rollback the transaction on failure.

**Review by BruceMacD — COMMENTED** [review]
I dont see any problems with this, but I would like @ssoroka to give it a look before approval where it's mostly his implementation.

File: internal/server/data/migrations_test.go
Comment (jmorganca) [inline internal/server/data/migrations_test.go:882]: Note: a migration test is needed / coming

File: internal/cmd/login.go
Comment (pdevine) [inline internal/cmd/login.go:189]: I _think_ this is fine. It's definitely better w/ the `for` loop here than the `goto`. I'm trying to think if we would ever break out of here on too many attempts (I mean, the user can always CTRL-C, so we have to handle the case anyway).

File: internal/cmd/login.go
Comment (pdevine) [inline internal/cmd/login.go:236]: is this bleed from @dnephin 's changes?

File: internal/cmd/login.go
Comment (pdevine) [inline internal/cmd/login.go:244]: I don't think you need this clause because the `strings.TrimPrefix()` part below should catch it anyway.

File: internal/server/data/deviceflowauthrequest.go
Comment (pdevine) [inline internal/server/data/deviceflowauthrequest.go:92]: Shouldn't this be `DELETE FROM device_flow_auth_requests`...? I thought we wanted to hard delete here so that the table doesn't keep filling up (i.e. soft delete in `DeleteDeviceFlowAuthRequest()` and then reap in this function).

File: internal/server/deviceflow.go
Comment (pdevine) [inline internal/server/deviceflow.go:72]: Why skip the access layer here? I think the handler will end up just passing through (i.e. the call here is unauthenticated), but we probably want to keep call through access to maintain the idiom (since everything else calls through it).

File: internal/server/deviceflow.go
Comment (pdevine) [inline internal/server/deviceflow.go:152]: small nit: this should probably be a constant

File: internal/server/deviceflow.go
Comment (pdevine) [inline internal/server/deviceflow.go:89]: this should probably be a constant (there's also "expired" above here)

File: internal/cmd/login.go
Comment (jmorganca) [reply internal/cmd/login.go:236]: No, `Login` and `loginToInfra` have been merged. Since device flow isn't a login method I had to restructure the logic around getting the login access key

File: internal/server/data/deviceflowauthrequest.go
Comment (jmorganca) [reply internal/server/data/deviceflowauthrequest.go:92]: Yes good catch! Thanks. This must have been from in a bad rebase

File: internal/cmd/login.go
Comment (jmorganca) [reply internal/cmd/login.go:244]: 💯  – this code was shuffled around and not added by this PR (unfortunately `git` didn't catch it). I went ahead and included that fix as part of this change!

File: internal/server/deviceflow.go
Comment (ssoroka) [inline internal/server/deviceflow.go:175]: this removes endpoint idempotency

**Review by ssoroka — CHANGES_REQUESTED** [review]
How does this solve the duplicate access key? Is that coming from somewhere else?

I think this is the wrong approach for reasons posted in comments.

File: internal/server/deviceflow.go
Comment (dnephin) [reply internal/server/deviceflow.go:138]: These 2 lines will make it so that the log line emitted for this request includes a `userID`. Without this the userID is not set because the request is not authenticated until the device code is approved.

File: internal/server/deviceflow.go
Comment (ssoroka) [inline internal/server/deviceflow.go:110]: device keys shouldn't be allowed to create more keys

File: internal/server/deviceflow.go
Comment (ssoroka) [reply internal/server/deviceflow.go:116]: clarification: ApproveDeviceFlow should fail if the approver's token is missing models.ScopeAllowCreateAccessKey

File: internal/server/deviceflow.go
Comment (ssoroka) [reply internal/server/deviceflow.go:178]: clarification: ApproveDeviceFlow should fail if the current key is missing models.ScopeAllowCreateAccessKey

File: internal/server/deviceflow.go
Comment (dnephin) [reply internal/server/deviceflow.go:110]: That seems like it's going to be a problem. Previously we were able to login from the CLI and create access keys. If device flow doesn't count as a login, then there's no way for someone to truly login via the CLI.

Comment (jmorganca) [reply internal/server/deviceflow.go:107]: I've added a new scope for access keys that allows them to approve device flow requests. The issue in the PR was not that device flow keys could create access keys (they are equivalent to login keys), but that they can continue to "renew" themselves by approving subsequent device flow requests.

File: internal/server/deviceflow.go
Comment (pdevine) [reply internal/server/deviceflow.go:113]: This is the same thing that happens in `server/authn/authn_method.go` inside of `Login()`, which is also an unauthenticated endpoint. The device flow is essentially the same thing as a login, only the _credential_ is the device code itself.

If we wanted to refactor this in the future, device flow really should just be another login method inside of `authn/`. I think the problem though is that code is a little confusing right now.

File: internal/server/models/access_key.go
Comment (dnephin) [inline internal/server/models/access_key.go:18]: These don't seem like they are really different scopes. Are we assigning these two scopes in all the same cases?

Device flow is effectively a way of getting an access key in the CLI. Can we continue to use `ScopeAllowCreateAccessKey` for approving of the request?

File: internal/server/deviceflow.go
Comment (pdevine) [reply internal/server/deviceflow.go]: On second thought, this should skip the access layer because it's essentially the same thing as the login handler (which also skips the access layer).

File: internal/server/deviceflow.go
Comment (pdevine) [reply internal/server/deviceflow.go]: We should probably be idempotent here, but ultimately once the polling code does the login, this record ends up being deleted and the approval will fail. So really there is a <= 5s window where this would be an issue.

File: internal/server/routes.go
Comment (pdevine) [reply internal/server/routes.go:135]: You would need to know this to get the provider/user (which is different than the old implementation).

File: internal/server/deviceflow.go
Comment (dnephin) [reply internal/server/deviceflow.go]: Definitely agree! I would go even further to say that all endpoints that do not require authorization should skip the access layer.

If the purpose of the access layer is "to notice when something is missing proper authorization", then a function in the `access` package that does not check authorization is actually defeating the purpose of this package.

File: internal/server/models/access_key.go
Comment (jmorganca) [reply internal/server/models/access_key.go:18]: They are close, but access keys resulting from logging with device flow should be able to:
1. Create an access key (since the user ran `infra login`), they should be able to run `infra keys add`
2. But not approve new device flows, since this would mean the user could continue to approve their own device flow requests

Comment (dnephin) [reply internal/server/models/access_key.go:18]: > But not approve new device flows, since this would mean the user could continue to approve their own device flow requests

Why is that a problem?

**Review by pdevine — APPROVED** [review]
lgtm

**Review by BruceMacD — APPROVED** [review]
Tested and validated that this revokes sessions on identity removal from the IDP

File: internal/cmd/login.go
Comment (dnephin) [inline internal/cmd/login.go:171]: Not a blocker for merging this, but do you know what causes the `NotFound`? Seems like maybe we should be handling that as `StatusUnauthorized` so that someone can't detect when a user exists or not.

File: internal/server/data/migrations.go
Comment (dnephin) [inline internal/server/data/migrations.go:1066]: What happens to existing device flow records when we run this migration? I think they will break because user_id and provider_id are null.

I guess since these are short lived and not many people are using it right now that it's ok, but still important to know. We should expect any active device flow request to fail when this is deployed to prod. Might be good to try and deploy during off-peak.

File: internal/server/deviceflow.go
Comment (dnephin) [inline internal/server/deviceflow.go:74]: This potentially allows someone to scan for device flow requests. Rate limiting should help mitigate that, but a distributed attack could still make lots of guesses. I guess since we delete it once approved it doesn't help them much.

Do we need to differentiate this error, or can we always return unauthorized?

File: internal/server/routes.go
Comment (dnephin) [inline internal/server/routes.go:135]: I think we could use the org_id from either the user or provider that we have IDs for, right? I'm not sure if that's necessary for this PR, but it seems like we don't necessarily need to get the org from the request domain.

However it would require relaxing our requirement that `data.GetIdentity` sets an orgID in the transaction metadata. We'd need an option to `GetIdentity` to tell it to skip the org requirement.

File: internal/server/deviceflow.go
Comment (dnephin) [inline internal/server/deviceflow.go:157]:
```suggestion
		// require the create access keys scope, because this approval allows the creation of an access key
```

File: internal/server/deviceflow.go
Comment (dnephin) [inline internal/server/deviceflow.go:158]: Just a note that we end up hiding all of these messages from the user: https://github.com/infrahq/infra/blob/8c6e5f361e561de77c6de330f6c77499d9d54140/internal/server/errors.go#L44-L47

We should find a way to allow us to send messages that we have written explicitly for these errors. I can create an issue for this.

File: internal/server/deviceflow_test.go
Comment (dnephin) [inline internal/server/deviceflow_test.go:141]: Is this TODO for a follow up? Or did you want to add it here?

**Review by dnephin — COMMENTED** [review]
LGTM! Some minor suggestions about the error code, but probably fine as follow ups.

There's a TODO in the test that might need to be removed or implemented.

File: internal/cmd/login.go
Comment (jmorganca) [reply internal/cmd/login.go:171]: This was moved from elsewhere, but `/api/login` should never return 404 for user discovery reasons and I don't think the endpoint ever does, so I'll remove that check.

File: internal/server/deviceflow.go
Comment (dnephin) [reply internal/server/deviceflow.go:158]: Created #3837

File: internal/server/deviceflow.go
Comment (jmorganca) [reply internal/server/deviceflow.go:74]: Great catch. Let's just return 401 here, as all public endpoints should.

File: internal/server/routes.go
Comment (jmorganca) [reply internal/server/routes.go:135]: Indeed. Went with this for now given `GetIdentity` requires the org id in the transaction

File: internal/server/data/migrations.go
Comment (jmorganca) [reply internal/server/data/migrations.go:1066]: This is a great catch. I'll add to the migration to expire existing requests immediately so any mid-device-flow users see a message and can restart the flow.

The alternative would be to populate `user_id` and `provider_id` from the access key, but these are short-lived and it would be better that we get the correct provider_id (which is not the case with the transient access key in the current model).

**Review by dnephin — APPROVED** [review]
Good idea to expire all the active requests. That should be a smoother upgrade experience.

**Pull request merged**
